### PR TITLE
[SPARK-45368][SQL] Remove scala2.12 compatibility logic for DoubleType, FloatType, Decimal

### PR DESCRIPTION
**Pull request by laglangyue**
### What changes were proposed in this pull request?

Remove scala2.12 compatibility logic for DoubleType, FloatType, Decimal

### Why are the changes needed?

Drop Scala 2.12 and make Scala 2.13 by default
https://issues.apache.org/jira/browse/SPARK-45368

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

test by ci

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by laglangyue** [comment]
As a beginner, I read some of the code here，I am amazed by the type design and submitted a PR.
Double and Float have removed the traits and directly inherited the traits of scala 2.13.
The decimal code is quite extensive, and I don't have a deep understanding，It seems to me that just need to remove comment，maybe it's incorrect.

File: sql/api/src/main/scala/org/apache/spark/sql/types/FloatType.scala [hunk 1]
```diff
@@ -43,26 +43,7 @@ class FloatType private() extends FractionalType {
 @Stable
 case object FloatType extends FloatType {
 
-  // Traits below copied from Scala 2.12; not present in 2.13
-  // TODO: SPARK-30011 revisit once Scala 2.12 support is dropped
```
Comment (HyukjinKwon) [inline sql/api/src/main/scala/org/apache/spark/sql/types/FloatType.scala:47]: Can we use this JIRA SPARK-30011?

**Comment by laglangyue** [comment]
I found that `asIntegral` was not used, PhysicalDecimalType.asIntegral was used directly, rather than through trait.

**Comment by laglangyue** [comment]
according to PR and `scala.math.Numeric.BigDecimalAsIfIntegral`
https://github.com/apache/spark/pull/26769/files
When we drop scala 2.12, all we just need to remove the comments.

**Comment by HyukjinKwon** [comment]
cc @srowen

**Comment by laglangyue** [comment]
Let me correct my previous description. 
Based on SPARK-30011,[SPARK-43110] [SQL] Move asIntegral to PhysicalDataType
Revised. Removed the implementation of asIntegral in double and float because they were not called.

File: sql/api/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 2]
```diff
@@ -681,8 +681,6 @@ object Decimal {
     override def toLong(x: Decimal): Long = x.toLong
     override def fromInt(x: Int): Decimal = new Decimal().set(x)
     override def compare(x: Decimal, y: Decimal): Int = x.compare(y)
-    // Added from Scala 2.13; don't override to work in 2.12
-    // TODO revisit once Scala 2.12 support is dropped
     def parseString(str: String): Option[Decimal] = Try(Decimal(str)).toOption
```
Comment (srowen) [inline sql/api/src/main/scala/org/apache/spark/sql/types/Decimal.scala:684]: I think this is meant to be `override` in Scala 2.13; does that work?

File: sql/api/src/main/scala/org/apache/spark/sql/types/DoubleType.scala [hunk 3]
```diff
@@ -40,34 +38,4 @@ class DoubleType private() extends FractionalType {
  * @since 1.3.0
  */
 @Stable
-case object DoubleType extends DoubleType {
-
-  // Traits below copied from Scala 2.12; not present in 2.13
```
Comment (srowen) [inline sql/api/src/main/scala/org/apache/spark/sql/types/DoubleType.scala:45]: In the previous PR that added these traits, were just these traits added for Scala 2.12/2.13? I am not sure if the object DoubleAsIfIntegral should be removed too

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/PhysicalDataType.scala [hunk 4]
```diff
@@ -92,7 +92,6 @@ object PhysicalNumericType {
 
 sealed abstract class PhysicalFractionalType extends PhysicalNumericType {
   private[sql] val fractional: Fractional[InternalType]
-  private[sql] val asIntegral: Integral[InternalType]
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/PhysicalDataType.scala:95]: What's the reason for removing this? I wasn't sure if this is related to Scala 2.12/2.13

File: sql/api/src/main/scala/org/apache/spark/sql/types/Decimal.scala (hunk 2 above)
Comment (laglangyue) [reply sql/api/src/main/scala/org/apache/spark/sql/types/Decimal.scala:684]: in scala2.12, Numberic not has this method, it ok, refer to `scala.math.Numeric.BigDecimalIsConflicted` in scala2.13

Comment (srowen) [reply sql/api/src/main/scala/org/apache/spark/sql/types/Decimal.scala:684]: Right, so shouldn't it be a `override` now?

Comment (laglangyue) [reply sql/api/src/main/scala/org/apache/spark/sql/types/Decimal.scala:684]: yes,I will.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/PhysicalDataType.scala (hunk 4 above)
Comment (laglangyue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/PhysicalDataType.scala:95]: same to above

**Comment by laglangyue** [comment]
@panbingkun @srowen 
Hi, hope to get your help. 
If Scala 2.12 is completely removed and only Scala 2.13 is used, is it necessary to remove the `asIntegral` method from the interface and double and float?Because I think this method has not been used, I hope to keep the same design as Scala 2.13.

File: sql/api/src/main/scala/org/apache/spark/sql/types/FloatType.scala [hunk 5]
```diff
@@ -41,37 +39,4 @@ class FloatType private() extends FractionalType {
...
-    def rem(x: Float, y: Float): Float = {
-      (BigDecimal(x.toDouble) remainder BigDecimal(y.toDouble)).floatValue
-    }
-  }
-
-  object FloatAsIfIntegral extends FloatAsIfIntegral {
-    override def compare(x: Float, y: Float): Int = java.lang.Float.compare(x, y)
-  }
-}
+case object FloatType extends FloatType
```
Comment (srowen) [inline sql/api/src/main/scala/org/apache/spark/sql/types/FloatType.scala:42]: Do we need to retain this object at all?

Comment (laglangyue) [reply sql/api/src/main/scala/org/apache/spark/sql/types/FloatType.scala:42]: sure，same as `org.apache.spark.sql.types.LongType`

**Comment by amaliujia** [comment]
> @amaliujia cc. shoudl `asIntegral` be removed for PhysicalFractionalType, float and double , I find the method not be used. if shoudle not removed,should we add this. `implicit object DoubleIsFractional extends DoubleIsFractional with Ordering.DoubleOrdering` because it is in scala2.12. thanks.

We keep the methods for the completeness, even though they may not be used. So for the completeness we should keep those.

**Comment by laglangyue** [comment]
> > @amaliujia cc. shoudl `asIntegral` be removed for PhysicalFractionalType, float and double , I find the method not be used. if shoudle not removed,should we add this. `implicit object DoubleIsFractional extends DoubleIsFractional with Ordering.DoubleOrdering` because it is in scala2.12. thanks.
> 
> We keep the methods for the completeness, even though they may not be used. So for the completeness we should keep those.

Get it , I have roll back.
By the way，do we need to redesign the double float based on `scala. path. Numeric. BigDecimalIsConflicted` in scala 2.13?It just seems like a slight difference

**Comment by laglangyue** [comment]
@srowen @amaliujia @HyukjinKwon PTAL

**Comment by srowen** [comment]
Merged to master

**Comment by panbingkun** [comment]
late LGTM.

**Comment by panbingkun** [comment]
It's strange that I haven't received any notification about this PR until today when I received the 'resolved' email on Jira.

**Review by beliefer — COMMENTED** [review]
LGTM later.

**Pull request closed without merging**